zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/user/DbUserState.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.engine.state.user;
+
+import static io.camunda.zeebe.util.buffer.BufferUtil.wrapString;
+
+import io.camunda.zeebe.db.ColumnFamily;
+import io.camunda.zeebe.db.DbKey;
+import io.camunda.zeebe.db.TransactionContext;
+import io.camunda.zeebe.db.ZeebeDb;
+import io.camunda.zeebe.db.impl.DbLong;
+import io.camunda.zeebe.db.impl.DbString;
+import io.camunda.zeebe.engine.state.immutable.UserState;
+import io.camunda.zeebe.engine.state.mutable.MutableUserState;
+import io.camunda.zeebe.protocol.ZbColumnFamilies;
+import io.camunda.zeebe.protocol.impl.record.value.user.UserRecord;
+import org.agrona.DirectBuffer;
+
+public class DbUserState implements UserState, MutableUserState {
+
+  private final PersistedUser persistedUser = new PersistedUser();
+
+  private final DbString username;
+  private final DbLong userKey;
+  private final ColumnFamily<DbString, DbLong> usernameToKeyColumnFamily;
+  private final ColumnFamily<DbKey, PersistedUser> keyToUserColumnFamily;
+
+  public DbUserState(
+      final ZeebeDb<ZbColumnFamilies> zeebeDb, final TransactionContext transactionContext) {
+    username = new DbString();
+    userKey = new DbLong();
+
+    usernameToKeyColumnFamily =
+        zeebeDb.createColumnFamily(
+            ZbColumnFamilies.USERNAME_BY_USER_KEY, transactionContext, username, userKey);
+
+    keyToUserColumnFamily =
+        zeebeDb.createColumnFamily(
+            ZbColumnFamilies.USERS, transactionContext, userKey, persistedUser);
+  }
```

Comment:
❌ Sorry, in the example I provided yesterday I forgot to add that we should use a Foreign Key in these ColumnFamilies. This ensures data consitency between the 2 Column Families. It's relatively easy to add:

```suggestion
  private final DbString username;
  private final DbLong userKey;
  private final DbForeignKey<DbLong> fkUserKey;
  private final ColumnFamily<DbString, DbForeignKey<DbLong>> usernameToKeyColumnFamily;
  private final ColumnFamily<DbKey, PersistedUser> keyToUserColumnFamily;
  
   public DbUserState(
       final ZeebeDb<ZbColumnFamilies> zeebeDb, final TransactionContext transactionContext) {
     username = new DbString();
     userKey = new DbLong();
     fkUserKey = new DbForeignKey<>(userKey, ZbColumnFamilies.USERS);

     usernameToKeyColumnFamily =
        zeebeDb.createColumnFamily(
            ZbColumnFamilies.USERNAME_BY_USER_KEY, transactionContext, username, fkUserKey);

     keyToUserColumnFamily =
        zeebeDb.createColumnFamily(
            ZbColumnFamilies.USERS, transactionContext, userKey, persistedUser);
```

So what we're saying with this is that key in the `usernameToKeyColumnFamily` is a foreign key. When we insert something into this ColumnFamily the key must exist in the `keyToUserColumnFamily`. Otherwise we're not allowed to insert it.

So this also means we need to swap the order in which we insert the data in the column families so we insert into the `keyToUserColumnFamily` first.


Sorry for not catching this sooner 🙇